src/storage/src/source/reclock.rs

```diff
@@ -7,931 +7,651 @@
-        .context("tried to invert empty frontier")?;
-
-    // If the since and the upper we are inverting are both == to 0, then
-    // we are either starting up for the first time, or we have a source that always
-    // starts at ts 0.
-    let zero = Antichain::from_elem(Timestamp::minimum());
-    if PartialOrder::less_equal(since, &zero)
-        && PartialOrder::less_equal(&upper_to_invert, &zero.borrow())
-    {
-        return Ok(OffsetAntichain::new());
-    }
-
-    // Assert we haven't compacted too far, and that we aren't (somehow) asking about the
-    // future.
-    if !PartialOrder::less_than(&since.borrow(), &upper_to_invert) {
-        return Err(anyhow::anyhow!(
-            "cannot invert {:?} because since ({:?}) is too great",
-            upper_to_invert,
-            since
-        ));
-    }
-    if !PartialOrder::less_equal(&upper_to_invert, &cur_upper.borrow()) {
-        return Err(anyhow::anyhow!(
-            "cannot invert {:?} because upper ({:?}) is too small",
-            upper_to_invert,
-            cur_upper,
-        ));
-    }
-
-    let mut source_upper = OffsetAntichain::with_capacity(remap_trace.len());
-    for pid in remap_trace.keys() {
-        let binding = partition_bindings(pid)
-            .take_while(|(ts, _)| ts < ts_to_invert)
-            .last();
-        if let Some((_, part_upper)) = binding {
-            source_upper.insert(pid.clone(), part_upper);
-        }
-    }
-    Ok(source_upper)
-}
-
 #[cfg(test)]
 mod tests {
     use super::*;
 
+    use std::collections::HashSet;
+    use std::sync::Arc;
     use std::time::Duration;
 
-    use mz_build_info::DUMMY_BUILD_INFO;
-    use mz_ore::now::{EpochMillis, SYSTEM_TIME};
+    use futures::Stream;
+    use itertools::Itertools;
+    use once_cell::sync::Lazy;
+    use timely::progress::Timestamp as _;
+    use tokio::sync::Mutex;
 
+    use mz_build_info::DUMMY_BUILD_INFO;
+    use mz_expr::PartitionId;
     use mz_ore::metrics::MetricsRegistry;
+    use mz_ore::now::{NowFn, SYSTEM_TIME};
+    use mz_persist_client::cache::PersistClientCache;
     use mz_persist_client::{PersistConfig, PersistLocation, ShardId};
+    use mz_repr::{GlobalId, Timestamp};
+    use mz_storage_client::controller::CollectionMetadata;
+    use mz_storage_client::types::sources::{MzOffset, SourceData};
+    use mz_timely_util::order::Partitioned;
 
     // 15 minutes
     static PERSIST_READER_LEASE_TIMEOUT_MS: Duration = Duration::from_secs(60 * 15);
 
-    fn persist_cache(now_fn: NowFn) -> Arc<Mutex<PersistClientCache>> {
+    static PERSIST_CACHE: Lazy<Arc<Mutex<PersistClientCache>>> = Lazy::new(|| {
         let mut persistcfg = PersistConfig::new(&DUMMY_BUILD_INFO, SYSTEM_TIME.clone());
 
         persistcfg.reader_lease_duration = PERSIST_READER_LEASE_TIMEOUT_MS;
-        persistcfg.now = now_fn;
+        let start = tokio::time::Instant::now();
+        persistcfg.now = NowFn::from(move || start.elapsed().as_millis().try_into().unwrap());
```

Comment:
can we not rely on actual time? or does the new impl actually make sure this works??

Reply:
I'm not too familiar with what this is doing, I just reverted a previous PR that tried to workaround the fact that the reclock operator was looking at the system clock directly instead of reacting to a clock stream.

Reply:
I don't think this matters at all anymore? what is the persist `now` even used for?